Mathlib/Topology/Algebra/Category/ProfiniteGrp.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Nailin Guan, Yuyang Zhao
+-/
+import Mathlib.Algebra.Category.Grp.FiniteGrp
+import Mathlib.Topology.Algebra.ContinuousMonoidHom
+import Mathlib.Topology.Category.Profinite.Basic
+
+
+/-!
+
+# Category of Profinite Groups
+
+We say `G` is a profinite group if it is a topological group which is compact and totally
+disconnected.
+
+## Main definitions and results
+
+* `ProfiniteGrp` is the category of profinite groups.
+
+* `ProfiniteGrp.pi` : The product of profinite groups is also a profinite group.
+
+-/
+
+suppress_compilation
+
+universe u
+
+open CategoryTheory Topology
+
+/--
+The category of profinite groups. A term of this type consists of a profinite
+set with a topological group structure.
+-/
+@[pp_with_univ]
+structure ProfiniteGrp where
+  /-- The underlying profinite topological space. -/
+  toProfinite : Profinite
+  /-- The group structure. -/
+  [group : Group toProfinite]
+  /-- The above data together form a topological group. -/
+  [topologicalGroup : TopologicalGroup toProfinite]
+
+/--
+The category of profinite additive groups. A term of this type consists of a profinite
+set with a topological additive group structure.
+-/
+@[pp_with_univ]
+structure ProfiniteAddGrp where
+  /-- The underlying profinite topological space. -/
+  toProfinite : Profinite
+  /-- The additive group structure. -/
+  [addGroup : AddGroup toProfinite]
+  /-- The above data together form a topological additive group. -/
+  [topologicalAddGroup : TopologicalAddGroup toProfinite]
+
+attribute [to_additive] ProfiniteGrp
+
+namespace ProfiniteGrp
+
+@[to_additive]
+instance : CoeSort ProfiniteGrp (Type u) where
+  coe G := G.toProfinite
+
+attribute [instance] group topologicalGroup
+    ProfiniteAddGrp.addGroup ProfiniteAddGrp.topologicalAddGroup
+
+@[to_additive]
+instance : Category ProfiniteGrp where
+  Hom A B := ContinuousMonoidHom A B
+  id A := ContinuousMonoidHom.id A
+  comp f g := ContinuousMonoidHom.comp g f
+
+@[to_additive]
+instance (G H : ProfiniteGrp) : FunLike (G ⟶ H) G H :=
+  inferInstanceAs <| FunLike (ContinuousMonoidHom G H) G H
+
+@[to_additive]
+instance (G H : ProfiniteGrp) : MonoidHomClass (G ⟶ H) G H :=
+  inferInstanceAs <| MonoidHomClass (ContinuousMonoidHom G H) G H
+
+@[to_additive]
+instance (G H : ProfiniteGrp) : ContinuousMapClass (G ⟶ H) G H :=
+  inferInstanceAs <| ContinuousMapClass (ContinuousMonoidHom G H) G H
+
+@[to_additive]
+instance : ConcreteCategory ProfiniteGrp where
+  forget :=
+  { obj := fun G => G
+    map := fun f => f }
+  forget_faithful :=
+    { map_injective := by
+        intro G H f g h
+        exact DFunLike.ext _ _ <| fun x => congr_fun h x }
+
+/-- Construct a term of `ProfiniteGrp` from a type endowed with the structure of a
+compact and totally disconnected (implies Hausdorff in topological group) topological group. -/
```

Comment:
Both in this doc and in the equivalent version for additive groups, what do you mean by "Hausdorff in topological groups"? That in the category of top. groups being compact and totally disconnected implies Hausdorff? If it is so, you can spell it more explicitely.

Reply:
Thanks!